sui/contract/sources/constant_product.move

```diff
@@ -89,154 +137,199 @@ module aux::constant_product {
         )
     }
 
-    /// Swap coin_x for coin_y (ETH-USDC would be +1 ETH & -2000 USDC)
-    public fun swap_exact_x_for_y<X, Y, W>(
-        self: &mut Pool<X, Y, W>, 
+    entry fun swap_exact_y_for_x_<X, Y>(
+        self: &mut Pool<X, Y>,
+        coin_y: Coin<Y>,
+        min_base_amount: u64,
+        ctx: &mut TxContext
+    ) {
+        transfer::transfer(
+            swap_exact_y_for_x(self, coin_y, min_base_amount, ctx),
+            tx_context::sender(ctx)
+        )
+    }
+
+    entry fun add_liquidity_<X, Y>(
+        self: &mut Pool<X, Y>,
+        coin_x: Coin<X>,
+        coin_y: Coin<Y>,
+        ctx: &mut TxContext
+    ) {
+        transfer::transfer(
+            add_liquidity(self, coin_x, coin_y, ctx),
+            tx_context::sender(ctx)
+        );
+    }
+
+    entry fun remove_liquidity_<X, Y>(
+        self: &mut Pool<X, Y>,
+        lp: Coin<LP<X, Y>>,
+        ctx: &mut TxContext
+    ) {
+        let (coin_x, coin_y) = remove_liquidity(self, lp, ctx);
+        let sender = tx_context::sender(ctx);
+
+        transfer::transfer(coin_x, sender);
+        transfer::transfer(coin_y, sender);
+    }
+
+    /********************/
+    /* Public functions */
+    /********************/
+
+    public fun create_pool<X, Y>(
+        pools: &mut Pools,
+        fee_bps: u64,
+        ctx: &mut TxContext
+    ) {
+        assert!(!vec_set::contains(&pools.pools, &type_name::get<Pool<X, Y>>()), EPoolAlreadyExists);
+        assert!((fee_bps as u128) <= MAX_BPS, EInvalidArg);
+        let supply_lp = balance::create_supply(LP {});
```

Comment:
Why not use `create_currency()` here?

Reply:
it enforces one-time witness which we can't satisfy because the lp coin is created not on init

Reply:
oh... but then LP coins aren't actually `Coin`s... so we can't create pools for them, etc.

Maybe we do need to deploy a new contract for each pool. That seems pretty limiting

Reply:
Oh wait nvm I see we can do `coin::from_balance()`

Reply:
Hm, but then there won't be a coin metadata... so how do we define its decimals, etc.?

Reply:
i'm not sure. for now, i've been stop-gapping missing functionality by emitting events, but that only works off-chain

Reply:
https://github.com/aux-exchange/aux-exchange/issues/335